src/nvtxw.rs

```diff
@@ -0,0 +1,385 @@
+                            short_name: ref short_name1,
+                            ..
+                        } = entry1
+                        {
+                            let mut x1: Vec<(String, String)> = Vec::new();
+                            for entry2 in slots2 {
+                                if let EntryInfo::Slot {
+                                    long_name: ref long_name2,
+                                    short_name: ref short_name2,
+                                    ..
+                                } = entry2
+                                {
+                                    let stream_name = format!("Legion {}", long_name2);
+                                    let domain_name = format!(
+                                        "Legion/{}/{}/{}",
+                                        short_name0, short_name1, short_name2
+                                    );
+
+                                    x1.push((stream_name, domain_name));
+                                }
+                            }
+
+                            x0.push(x1);
+                        }
+                    }
+
+                    panels.push(x0);
+                }
+            }
+        }
+
+        let mut entries: HashMap<u64, EntryContents> = HashMap::new();
+        while self.data_source.outstanding_requests() > 0 {
+            /*
+            for tile in self.data_source.get_summary_tiles() {
+            }
+            */
+            for tile in self.data_source.get_slot_tiles() {
+                let entry_id = tile.entry_id;
+                let items = tile.data.items;
+                for item_group in items {
+                    for item in item_group {
+                        let item_uid = item.item_uid;
+                        let time_start = item.interval.start;
+                        let time_stop = item.interval.stop;
+                        let color = item.color;
+
+                        let key = item_uid.0;
+
+                        match entries.get_mut(&key) {
+                            Some(ref mut e) => {
+                                assert!(e.entry_id == entry_id);
+
+                                e.entry_id = entry_id.clone();
+                                e.color = color;
+                            }
+                            _ => {
+                                let e = EntryContents {
+                                    entry_id: entry_id.clone(),
+                                    time_start,
+                                    time_stop,
+                                    color,
+                                    title: String::new(),
+                                };
+
+                                entries.insert(key, e);
+                            }
+                        }
+                    }
+                }
+            }
+            for tile in self.data_source.get_slot_meta_tiles() {
```

Comment:
So far, I've struggled with implementing batching. I have the code restructured so that I can follow the paradigm used in archive_data.rs, but I think the logic I'm using to match is not correct. In the current push, I've kept the matching logic the same as before, as well as waiting on all tiles rather than batching, so that the output is the also same as before, but I think even that output is not correct, only containing about 2500 events for the circuit example rather than 7500, because the same tiles are matched more than once.

Do you see what I'm doing wrong? I suppose checking EntryUID is not enough--perhaps I should add tile ID comparisons too. I'm also not making use of any ordering so far.

Reply:
I think what you're missing is that there can be multiple `Item`s associated with an `ItemUID`. This happens when a task blocks and goes into a waiting state. So for example, you might have:

 1. From timestamp T0 to T1: task is running
 2. At T1: task blocks
 3. From T1 to T2: task is blocked
 4. At T2: task becomes ready
 5. From T2 to T3: task is ready but not running
 6. At T3: task starts running
 7. From T3 to T4: task is running

In this example you would end up with 4 `Item`s:

 1. `Item` from T0 to T1 in running state
 2. `Item` from T1 to T2 in blocked state
 3. `Item` from T2 to T3 in ready state
 4. `Item` from T3 to T4 in running state
 
 Because they're all the same task, they have the same `ItemUID`.
 
 Here are the invariants I think you can rely on, as long as you always ask for one tile spanning the entire profile with `full=true`. All of the `Item`s with a given `ItemUID` will be guaranteed to be:

 1. On the same row (what you call `item_group` in your code)
 2. Consecutive in array order (i.e., no other items will be interleaved with them)
 3. Consecutive in time (i.e., the stop of one item will always be identical to the start of the next)
 4. The `Item`s and `ItemMeta`s will always appear in the same order and positions in their respective tiles

Let me know if you have any questions.

Reply:
By the way, some of the invariants above would have been violated by previous Legion versions, but if you update to the newest `master` they should all be preserved now.

Reply:
While we're here, I wanted to remind you about this comment I made last time:

https://github.com/StanfordLegion/prof-viewer/pull/54#pullrequestreview-2006506214

Fundamentally, you hit the issue with `ItemUID`s because you're slurping the entire profile into memory and matching `Item`s across tiles. There are two issues. The first is with managing the invariants of multiple items as you noted. The second is that it uses memory proportional to the size of the profile.

I've seen cases where this simply won't fit into memory. Mike Bauer showed me one the other day from cuNumeric. So this is simply a no-go for use cases that I'm pretty sure you'll care about. But since it should also simplify the implementation to do things are tile granularity, I think it would just make the code nicer to deal with overall.

Reply:
I believe I have finally figured out how to implement the batching you've requested. Please let me know if the code is indeed doing what you intended.

Reply:
Structurally, I think this should work. I provided a couple of suggestions in the code: one should simply the structure further, and the other is a code style issue that would make the code more straightforward.

I have not re-reviewed the NVTXW code specifically. Would you like me to do so now?

Reply:
Thank you for your review, I see how using a pair of `Option`s could simplify the code further.

> I have not re-reviewed the NVTXW code specifically. Would you like me to do so now?

Yes please, the NVTXW crate code is ready for further review.